wicket-util/src/main/java/org/apache/wicket/util/crypt/AESCrypt.java

```diff
@@ -0,0 +1,112 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.wicket.util.crypt;
+
+import javax.crypto.BadPaddingException;
+import javax.crypto.Cipher;
+import javax.crypto.IllegalBlockSizeException;
+import javax.crypto.SecretKey;
+import javax.crypto.spec.IvParameterSpec;
+
+/**
+ * AES based {@link ICrypt} encrypt and decrypt strings such as passwords or URL segments.
+ * Based on http://stackoverflow.com/a/992413
+ * 
+ * @see ICrypt
+ */
+class AESCrypt extends AbstractJceCrypt
+{
+
+	private final SecretKey secretKey;
+	private final String algorithm;
+	private final int ivSize;
+
+	
+	/**
+	 * Constructor
+	 * 
+	 * @param secretKey
+	 *              The {@link SecretKey} to use to initialize the {@link Cipher}.
+	 */
+	public AESCrypt(SecretKey secretKey)
+	{
+		this(secretKey, "AES/CBC/PKCS5Padding", 16);
+	}
+
+	/**
+	 * Constructor
+	 * 
+	 * @param secretKey
+	 *              The {@link SecretKey} to use to initialize the {@link Cipher}.
+	 * @param algorithm
+	 *              The cipher algorithm to use, for example "AES/CBC/PKCS5Padding".
+	 * @param ivSize
+	 *              The size of the Initialization Vector to use with the cipher.
```

Comment:
The ivSize is not needed. It is always equal to the blocksize of the cipher. So you can simply use cipher.getBlockSize() to get it.

Reply:
Right, but during decipher phase you need ivSize before creating the cipher reading the first ivSize of encrypted data. So I would leave this constructor parameter

Reply:
But that's only because the lookup and unit of the Cipher are in `buildCipher`. IMHO that method does not really make things easier to understand. I think the code would be better with that method unlined. Then you can bring the Cipher lookup to the front and use that to get the block size.

This is just my opinion. If you do decide to keep the method, then rename the field to `blockSize`, as that is what it is.

Reply:
Sorry but you are loosing me here. How could you build a Cipher in DECRYPT_MODE without already having the expected IV (not just the size of it)?  Is it possible to write a version of your method _decrypt_ posted above without knowing the fixed size of IV?

Reply:
You get the Cipher with just `Cipher.getInstance(alg)`. There you can get the block size from. You only need the iv for the call to `init`. I'm typing this on my phone, so posting code is a bit difficult. What about this?

```public byte[] decrypt(SecureRandom rnd, SecretKey key, byte[] cipherInput) throws GeneralSecurityException {
    Cipher cipher = Cipher.getInstance("AES/CBC/PKCS5Padding");
    int blocksize = cipher.getBlockSize();
    byte[] iv = new byte[blocksize];
    byte[] ciphertext = new byte[cipherInput.length - blocksize];
    System.arraycopy(cipherInput, 0, iv, 0, iv.length);
    System.arraycopy(cipherInput, blocksize, ciphertext, 0, ciphertext.length);

    
    cipher.init(Cipher.DECRYPT_MODE, key, new IvParameterSpec(iv), rnd);
    return cipher.doFinal(ciphertext);
}```

Reply:
Thank you, now I see... :-)